libraries/standard/mqtt/src/mqtt.c

```diff
@@ -130,6 +234,435 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+
+    /* If keep alive interval is 0, it is disabled. */
+    if( ( keepAliveMs != 0U ) && 
+        ( calculateElapsedTime( now, pContext->lastPacketTime ) > keepAliveMs ) )
+    {
+        if( pContext->waitingForPingResp )
+        {
+            /* Has time expired? */
+            if( calculateElapsedTime( now, pContext->pingReqSendTimeMs ) > 
+                pContext->pingRespTimeoutMs )
+            {
+                status = MQTTKeepAliveTimeout;
+            }
+        }
+        else
+        {
+            status = MQTT_Ping( pContext );
+        }
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t handleIncomingPublish( MQTTContext_t * const pContext,
+                                           MQTTPacketInfo_t * pIncomingPacket )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t publishRecordState = MQTTStateNull;
+    uint16_t packetIdentifier;
+    MQTTPublishInfo_t publishInfo;
+
+    assert( pContext != NULL );
+    assert( pIncomingPacket != NULL );
+
+    status = MQTT_DeserializePublish( pIncomingPacket, &packetIdentifier, &publishInfo );
+    LogInfoWithArgs( "De-serialized incoming PUBLISH packet: DeserializerResult=%d", status );
+
+    if( status == MQTTSuccess )
+    {
+        publishRecordState = MQTT_UpdateStatePublish( pContext,
+                                                      packetIdentifier,
+                                                      MQTT_RECEIVE,
+                                                      publishInfo.qos );
+        LogInfoWithArgs( "State record updated. New state=%d.",
+                         publishRecordState );
+
+        /* Send PUBACK or PUBREC if necessary. */
+        status = sendPublishAcks( pContext,
+                                  packetIdentifier,
+                                  &publishRecordState );
+    }
+
+    if( status == MQTTSuccess )
+    {
+        /* TODO: Should the publish info be passed instead? */
+        pContext->callbacks.appCallback( pContext, pIncomingPacket, packetIdentifier );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t handleIncomingAck( MQTTContext_t * const pContext,
+                                       MQTTPacketInfo_t * pIncomingPacket )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t publishRecordState = MQTTStateNull;
+    uint16_t packetIdentifier;
+    /* Need a dummy variable for MQTT_DeserializeAck(). */
+    bool sessionPresent = false;
+    MQTTPubAckType_t ackType;
+
+    assert( pContext != NULL );
+    assert( pIncomingPacket != NULL );
+
+    switch( pIncomingPacket->type )
+    {
+        case MQTT_PACKET_TYPE_PUBACK:
+        case MQTT_PACKET_TYPE_PUBREC:
+        case MQTT_PACKET_TYPE_PUBREL:
+        case MQTT_PACKET_TYPE_PUBCOMP:
+            ackType = getAckFromPacketType( pIncomingPacket->type );
```

Comment:
minor:can this be moved inside if (status == MQTTSuccess)? Doesn't seems like it is needed if there is an error.

Reply:
Yes it can, the reason I didn't put it there was because I didn't want to increase the complexity. I can move it if you think it should be done.

Reply:
I am ok with either apporach.